### PR TITLE
Run hb_report under hacluster

### PR DESCRIPTION
**Pull request by liangxin1300**
## Problem
It didn't work if run **hb_report** or **crm cluster copy** under **hacluster** user, which was required from Hawk

## Solution

- Change function `bootstrap.configure_local_ssh_key` and `bootstrap.swap_public_ssh_key` to accept non-root user parameter; That is, for common user, bootstrap process also setup password-less among cluster nodes
- Assign `/usr/bin/sh` to hacluster's login shell, otherwise, it's impossible to run command remotely
- On interactive mode, behavior about changing hacluster's login shell and setup ssh access will be asked for confirm
- For `runtime` existing cluster, need to use `ssh` stage to setup password-less access for hacluster
  * on init node, run `crm cluster init ssh -y`
  * on join node, run `crm cluster join ssh -c <init node> -y`
  
- For `runtime` existing cluster, `/etc/corosync/corosync.conf` and `/etc/sysconfig/sbd` might has 400 mod, which will cause `Permission denied` exceptions when running hb_report under hacluster, #747 try to resolve this, but not works for `runtime` existing cluster, to avoid that, before running hb_report under hacluster:
  * Run `chmod 644 /etc/corosync/corosync.conf`
  * Run `chmod 644 /etc/sysconfig/sbd`

Test rpm link: https://build.opensuse.org/package/show/home:XinLiang:branches:crmsh_testing7/crmsh

### CODE REVIEW

**Review by MalloZup — COMMENTED** [review]
seems good to me but not familiar with crmsh codebase

File: crmsh/bootstrap.py
Comment (arbulu89) [inline crmsh/bootstrap.py:1148]: I wouldn't return a 3 elements list here. This is a potential issue if for some reason you check the return data, as you will need to change all the places where the method is used.
Better to return a dictionary or specific object, so you can use just the needed elements

File: crmsh/bootstrap.py
Comment (arbulu89) [inline crmsh/bootstrap.py:1136]: This `["root", "hacluster"]` list looks something that should go in a constant, as it is repeated later

File: crmsh/bootstrap.py
Comment (arbulu89) [inline crmsh/bootstrap.py:1180]: Maybe adding the next code in a method is a good idea:
```
if user != "root":
  cmd = utils.add_su(cmd, user)
```
It is reused in different places

**Review by arbulu89 — COMMENTED** [review]
@liangxin1300 Added some comments about the code.
About the purpose of the changes, I cannot really judge without knowing what are the consequences of changing the `hacluster` user bash option

File: crmsh/bootstrap.py
Comment (MalloZup) [reply crmsh/bootstrap.py:1148]: yop valid point

Comment (liangxin1300) [reply crmsh/bootstrap.py]: Thanks @arbulu89 , already changed!

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Already put it into `add_su` function

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Changed

**Review by MalloZup — APPROVED** [review]

**Review by gao-yan — CHANGES_REQUESTED** [review]
It generally looks good to me. Thanks for the nice work, @liangxin1300!

Only one nitpick.

**Comment by gao-yan** [comment]
> * For `runtime` existing cluster, need to use `ssh` stage to setup password-less access for hacluster
>   
>   * on init node, run `crm cluster init ssh -y`
>   * on join node, run `crm cluster join ssh -c <init node> -y`

Probably it could be a further improvement. To make it as friendly as possible, given that there's an existing cluster, it'd be great to only require running such a `ssh` stage on any of nodes, which takes care of all the nodes.

> * For `runtime` existing cluster, `/etc/corosync/corosync.conf` and `/etc/sysconfig/sbd` might has 400 mod, which will cause `Permission denied` exceptions when running hb_report under hacluster, #747 try to resolve this, but not works for `runtime` existing cluster, to avoid that, before running hb_report under hacluster:
>   
>   * Run `chmod 644 /etc/corosync/corosync.conf`
>   * Run `chmod 644 /etc/sysconfig/sbd`
> 

Also to make it as easy as possible for users, I'm thinking probably `ssh` stage running for an existing cluster should as well fix the permissions of any theses existing files?

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py:1176]: Since 
```
ls -al /usr/bin/sh
lrwxrwxrwx 1 root root 9 Sep 10  2019 /usr/bin/sh -> /bin/bash
```
```
# ls -al /bin/sh
lrwxrwxrwx 1 root root 4 Sep 10  2019 /bin/sh -> bash
```
I decide to change this as `/bin/bash`

**Comment by liangxin1300** [comment]
> > * For `runtime` existing cluster, need to use `ssh` stage to setup password-less access for hacluster
> >   
> >   * on init node, run `crm cluster init ssh -y`
> >   * on join node, run `crm cluster join ssh -c <init node> -y`
> 
> Probably it could be a further improvement. To make it as friendly as possible, given that there's an existing cluster, it'd be great to only require running such an `ssh` stage on any of the nodes, which takes care of all the nodes.

Good suggestion!  Thanks @gao-yan 
> 
> > * For `runtime` existing cluster, `/etc/corosync/corosync.conf` and `/etc/sysconfig/sbd` might has 400 mod, which will cause `Permission denied` exceptions when running hb_report under hacluster, #747 try to resolve this, but not works for `runtime` existing cluster, to avoid that, before running hb_report under hacluster:
> >   
> >   * Run `chmod 644 /etc/corosync/corosync.conf`
> >   * Run `chmod 644 /etc/sysconfig/sbd`
> 
> Also to make it as easy as possible for users, I'm thinking probably `ssh` stage running for an existing cluster should as well fix the permissions of any theses existing files?

I agree that will be more convenient for this whole issue.
But ssh key and the permissions changing are two kinds of things for crmsh, is that proper to put them together in `ssh` stage?

**Comment by gao-yan** [comment]
> > > * For `runtime` existing cluster, `/etc/corosync/corosync.conf` and `/etc/sysconfig/sbd` might has 400 mod, which will cause `Permission denied` exceptions when running hb_report under hacluster, #747 try to resolve this, but not works for `runtime` existing cluster, to avoid that, before running hb_report under hacluster:
> > >   
> > >   * Run `chmod 644 /etc/corosync/corosync.conf`
> > >   * Run `chmod 644 /etc/sysconfig/sbd`
> > 
> > 
> > Also to make it as easy as possible for users, I'm thinking probably `ssh` stage running for an existing cluster should as well fix the permissions of any theses existing files?
> 
> I agree that will be more convenient for this whole issue.
> But ssh key and the permissions changing are two kinds of things for crmsh, is that proper to put them together in `ssh` stage?

Technically, no :-)  As we can see, the "breaking" changes of hawk will bring  users  much inconvenience requiring them to re-run ssh stage.  I was thinking combing this into ssh stage would make their lives a bit easier :-)  But of course, we could either add this step also into update note / TID ...

**Comment by liangxin1300** [comment]
> > > > * For `runtime` existing cluster, `/etc/corosync/corosync.conf` and `/etc/sysconfig/sbd` might has 400 mod, which will cause `Permission denied` exceptions when running hb_report under hacluster, #747 try to resolve this, but not works for `runtime` existing cluster, to avoid that, before running hb_report under hacluster:
> > > >   
> > > >   * Run `chmod 644 /etc/corosync/corosync.conf`
> > > >   * Run `chmod 644 /etc/sysconfig/sbd`
> > > 
> > > 
> > > Also to make it as easy as possible for users, I'm thinking probably `ssh` stage running for an existing cluster should as well fix the permissions of any theses existing files?
> > 
> > 
> > I agree that will be more convenient for this whole issue.
> > But ssh key and the permissions changing are two kinds of things for crmsh, is that proper to put them together in `ssh` stage?
> 
> Technically, no :-) As we can see, the "breaking" changes of hawk will bring users much inconvenience requiring them to re-run ssh stage. I was thinking combing this into ssh stage would make their lives a bit easier :-) But of course, we could either add this step also into update note / TID ...

@arbulu89 , How do you think about this?

**Comment by liangxin1300** [comment]
> > > * For `runtime` existing cluster, need to use `ssh` stage to setup password-less access for hacluster
> > >   
> > >   * on init node, run `crm cluster init ssh -y`
> > >   * on join node, run `crm cluster join ssh -c <init node> -y`
> > 
> > 
> > Probably it could be a further improvement. To make it as friendly as possible, given that there's an existing cluster, it'd be great to only require running such an `ssh` stage on any of the nodes, which takes care of all the nodes.
> 
> Good suggestion! Thanks @gao-yan

@gao-yan After estimating, I think this change might be big and might break compatibility and bring regression.
So I suggest keep this unchanged and try to clarify these in `update note`

File: crmsh/bootstrap.py
Comment (arbulu89) [inline crmsh/bootstrap.py:1161]: I think this `configure_local_ssh_key` method has become too big and complex. 
We could split the logic in different submethods

**Review by arbulu89 — APPROVED** [review]
@liangxin1300 
I see correct the code changes. But I'm talking about python code itself.
As commented, I don't have that much context about the change, so I would wait until @gao-yan approval to move forward

File: crmsh/bootstrap.py
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Thanks @arbulu89 
I moved the change shell codes into `change_user_shell` function
And I will adjust UT codes later

**Review by gao-yan — APPROVED** [review]

**Comment by liangxin1300** [comment]
Loop #1033 here

**Pull request merged**
